sources/curriculum/0_before_you_start/workflows-and-pipelines-detailed/README.md

```diff
@@ -0,0 +1,194 @@
+# Pipelines and Project Workflow
+
+## Motivation
+
+Most data-science projects have the same set of tasks:
+
+1. **ETL**: extracting data from its source, transforming it, then loading it into a database.
+2. **Pre-process data**: This might include imputing missing values and choosing the training and testing sets.
+3. **Train the model(s)**: You can try different algorithms, features, and
+so on.
+4. Assess performance on the test set: Using an appropriate
+metric (e.g. $Precision@k$, $Recall$, $AUC$), examine the performance of your model "out of
+sample."
+5. Think of new things to try. Repeat steps 1 through 4 as appropriate.
+
+If the code base is not structured and named well, you might be struggling
+to remember the details of each step once you have built a few models. 
+What features did you use for each?
+What training and testing split?  What hyperparameters?
+
+Your code might be getting messy too. Did you overwrite the code for
+the previous model? Maybe you copied, pasted, and edited code from an
+earlier model. Can you still read what's there? It can quickly become
+a hodgepodge that requires heroics to decipher.
+
+In this session, we will introduce a workflow that can avoid (or at least reduce) these problems.
+
+# Data pipelines
+
+It is helpful to structure the data into multiple layers.
+
+##### Raw
+The data we receive from the partners and external sources is the raw data. 
+Raw data is immutable. Quoting from the popular workflow package [Data Science Cookiecutter](https://drivendata.github.io/cookiecutter-data-science/#data-is-immutable):
+
+Don't ever edit your raw data, especially not manually, and especially not in Excel. 
+Don't overwrite your raw data. 
+Don't save multiple versions of the raw data. 
+Treat the data (and its format) as immutable. 
+The code you write should move the raw data through a pipeline to your final analysis. 
+You shouldn't have to run all of the steps every time you want to make a new figure (see Analysis is a DAG), but anyone should be able to reproduce the final products with only the code in src and the data in data/raw.
+
+
+##### Intermediate
+If the raw data is messy, it is advisable to create an intermediate layer that consists of tidy copies of the raw data.
+Typical situations where this is useful are:
+- Data is received in multiple different file types 
+- Data fields are not typed (e.g. csv files, excel) or poorly typed (dates as strings, inconsistent date formats)
+- Column names are unclear, have spaces, special characters, or there are no column names 
+
+The transformations from raw to intermediate should be limited to fix the issues mentioned above. 
+We should not combine different data sets or create calculated fields.
+This is reserved for the next layer.
+
+Typical file formats for the intermediate layer are a data base (e.g. `postgres`) or `parquet` files.
```

Comment:
Technically, `postgres` are compose by files (everything is a file) but no one thinks in databases as files. Could you re-redact this? 

Also, if we are talking about `parquet`... are we planning to explain advantages of this file format somewhere? Are we planning to use it? The interface with plain `python` depends in arrow. Why not talking about arrow then? Do we need to have this discussion?

Reply:
Agreed. I improved wording.
On the second question, I think it makes sense to discuss parquet files since it's the most common storage format as far as I know. But we can introduce it a bit later in the programme.

Reply:
Agree, but in the other side, in most of the Big data projects (where a database is not the best choice, you know above ~ 6 Tb) you will be ending storing the files in different formats depending on the stage and the transformation: ORC will be faster for some joins, parquet will be useful for some columnar calculations, AVRO will be good for final storage (given that is resilient for corruption), JSON for sending data to the  browser, etc.

The same applies for DBs.

So even in one project you will potentially have different formats, in the same pipeline.

Reply:
Agreed. I have pointed out the need for a variety of storage formats in the reporting layer.